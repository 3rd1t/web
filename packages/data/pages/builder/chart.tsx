import React from "react"

/* eslint-disable-next-line */
export interface ChartProps {}

export const Chart = (props: ChartProps) => {
  return (
    <div>
      <h1>Welcome to chart!</h1>
    </div>
  )
}

export default Chart
