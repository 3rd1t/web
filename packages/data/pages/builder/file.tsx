import React, { useState, useEffect } from "react"
import { MultiColumn } from "@perfolio/shared/nav"
import { MultistepForm, Table, Button } from "@perfolio/shared/components"
import { AnimatePresence, AnimateSharedLayout, motion } from "framer-motion"

export interface FileProps {
  factorModels: { name: string; description: string; factors: { name: string; description: string }[] }[]
  regions: string[]
  currencies: string[]
  intervals: string[]
}

export const File = (props: FileProps) => {
  const [model, setModel] = useState<number>(-1)
  const [factor, setFactor] = useState<number>(-1)
  const [region, setRegion] = useState<number>(-1)
  const [currency, setCurrency] = useState<number>(-1)
  const [interval, setInterval] = useState<number>(-1)

  useEffect(() => {
    setFactor(-1)
  }, [model])

  const steps: {
    name: string
    description: string
    choices: string[]
    value: number
    setValue: React.Dispatch<React.SetStateAction<number>>
  }[] = [
    {
      name: "Model",
      description: "Quisque ultrices odio ut tellus congue, scelerisque egestas augue accumsan.",
      choices: props.factorModels.map((m) => m.description),
      value: model,
      setValue: setModel,
    },
    {
      name: "Factor",
      description: "Quisque ultrices odio ut tellus congue, scelerisque egestas augue accumsan.",
      choices:
        model >= 0
          ? props.factorModels.find((f) => f.name === props.factorModels[model].name).factors.map((f) => f.description)
          : [],
      value: factor,
      setValue: setFactor,
    },
    {
      name: "Region",
      description: "Quisque ultrices odio ut tellus congue, scelerisque egestas augue accumsan.",
      choices: props.regions,
      value: region,
      setValue: setRegion,
    },
    {
      name: "Currency",
      description: "Quisque ultrices odio ut tellus congue, scelerisque egestas augue accumsan.",
      choices: props.currencies,
      value: currency,
      setValue: setCurrency,
    },
    {
      name: "Interval",
      description: "Quisque ultrices odio ut tellus congue, scelerisque egestas augue accumsan.",
      choices: props.intervals,
      value: interval,
      setValue: setInterval,
    },
  ]

  const [cells, setCells] = useState([])
  const [complete, setComplete] = useState(false)
  // Checks if form is complete and sets table content
  useEffect(() => {
    if (model >= 0 && factor >= 0 && region >= 0 && currency >= 0 && interval >= 0) {
      setComplete(true)
    } else {
      setComplete(false)
    }
  }, [model, factor, region, currency, interval])

  const path = () => {
    return [
      props.factorModels[model].name,
      props.factorModels[model].factors[factor].name,
      props.regions[region],
      props.currencies[currency],
      props.intervals[interval],
    ].join("/")
  }

  const columnNames = (): string[] => {
    if (model < 0 || factor < 0) {
      return []
    }

    const selectedFactor = props.factorModels[model].factors[factor].name
    if (selectedFactor.toLowerCase() === "all") {
      return ["Date", ...props.factorModels[model].factors.map((f) => f.name)]
    }
    return ["Date", props.factorModels[model].factors[factor].name]
  }

  return (
    <div>
      <AnimateSharedLayout>
        <MultiColumn breadcrumbs={["data", "builder", "file"]}>
          <div className="p-8 space-y-4">
            <h3 className="text-lg font-semibold text-gray-800">Build your custom file</h3>
            <MultistepForm steps={steps} />

            <AnimatePresence>
              {complete ? (
                <>
                  <motion.div
                    initial={{ opacity: 0 }}
                    animate={{ opacity: 1 }}
                    exit={{ opacity: 0 }}
                    transition={{ damping: 300 }}
                  >
                    <Table columnNames={columnNames()} cells={cells}></Table>
                  </motion.div>

                  <motion.div
                    layout
                    className="mt-16 text-center"
                    initial={{ opacity: 0 }}
                    animate={{ opacity: 1 }}
                    exit={{ opacity: 0 }}
                    transition={{ damping: 300 }}
                  >
                    <Button
                      onClick={() => console.log("DONE")}
                      label="Download"
                      bgColor="bg-carbon-900"
                      textColor="text-gray-100"
                      iconRight={
                        <svg
                          className="w-6 h-6"
                          fill="none"
                          stroke="currentColor"
                          viewBox="0 0 24 24"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            strokeLinecap="round"
                            strokeLinejoin="round"
                            strokeWidth={2}
                            d="M12 10v6m0 0l-3-3m3 3l3-3m2 8H7a2 2 0 01-2-2V5a2 2 0 012-2h5.586a1 1 0 01.707.293l5.414 5.414a1 1 0 01.293.707V19a2 2 0 01-2 2z"
                          />
                        </svg>
                      }
                    ></Button>
                  </motion.div>
                </>
              ) : null}
            </AnimatePresence>
          </div>
        </MultiColumn>
      </AnimateSharedLayout>
    </div>
  )
}

export default File

export async function getStaticProps() {
  return {
    props: {
      factorModels: [
        {
          name: "3factor",
          description: "3 Factor model by Fama & French 1993",
          factors: [
            {
              name: "XXX",
              description: "maybe just spell it out",
            },
            {
              name: "XXX",
              description: "maybe just spell it out",
            },
            {
              name: "XXX",
              description: "maybe just spell it out",
            },
            {
              name: "all",
              description: "All factors",
            },
          ],
        },
        {
          name: "4factor",
          description: "4 Factor model by Fama & French 1993",

          factors: [
            {
              name: "XXX",
              description: "maybe just spell it out",
            },
            {
              name: "XXX",
              description: "maybe just spell it out",
            },
            {
              name: "XXX",
              description: "maybe just spell it out",
            },
            {
              name: "XXX",
              description: "maybe just spell it out",
            },
            {
              name: "all",
              description: "All factors",
            },
          ],
        },
        {
          name: "5factor",
          description: "5 Factor model by Fama & French 1993",

          factors: [
            {
              name: "XXX",
              description: "maybe just spell it out",
            },
            {
              name: "XXX",
              description: "maybe just spell it out",
            },
            {
              name: "XXX",
              description: "maybe just spell it out",
            },
            {
              name: "XXX",
              description: "maybe just spell it out",
            },
            {
              name: "XXX",
              description: "maybe just spell it out",
            },

            {
              name: "all",
              description: "All factors",
            },
          ],
        },
        {
          name: "6factor",
          description: "6 Factor model by Fama & French 1993",

          factors: [
            {
              name: "mom",
              description: "Momentum",
            },
            {
              name: "XXX",
              description: "maybe just spell it out",
            },
            {
              name: "XXX",
              description: "maybe just spell it out",
            },
            {
              name: "XXX",
              description: "maybe just spell it out",
            },
            {
              name: "XXX",
              description: "maybe just spell it out",
            },
            {
              name: "XXX",
              description: "maybe just spell it out",
            },

            {
              key: "all",
              description: "All factors",
            },
          ],
        },
      ],
      regions: [
        "USA",
        "Developed",
        "Developed_ex_US",
        "Europe",
        "Japan",
        "Asia_Pacific_ex_Japan",
        "North_America",
        "Emerging",
      ],

      currencies: [
        "EUR",
        "JPY",
        "GBP",
        "CHF",
        "RUB",
        "AUD",
        "BRL",
        "CAD",
        "CNY",
        "INR",
        "DKK",
        "NZD",
        "NOK",
        "SEK",
        "PLN",
        "ILS",
        "KRW",
        "TRY",
      ],

      intervals: ["daily", "monthly", "annual"],
    },
  }
}
