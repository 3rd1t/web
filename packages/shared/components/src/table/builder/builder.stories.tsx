import React from "react"
import { SimpleTable } from "./builder"
export default {
  title: "components/table/builder",
}

export const simpleTable = () => {
  const rows = [
    {
      period: "2019",
      mktrf: "0.33490000",
      smb: "-0.06050000",
      hml: "-0.12380000",
    },
    {
      period: "2018",
      mktrf: "-0.00330000",
      smb: "-0.03720000",
      hml: "-0.09630000",
    },
    {
      period: "2017",
      mktrf: "0.08100000",
      smb: "-0.04230000",
      hml: "-0.12250000",
    },
    {
      period: "2016",
      mktrf: "0.17440000",
      smb: "0.06830000",
      hml: "0.23620000",
    },
    {
      period: "2015",
      mktrf: "0.11890000",
      smb: "-0.04520000",
      hml: "-0.10770000",
    },
    {
      period: "2014",
      mktrf: "0.26530000",
      smb: "-0.09140000",
      hml: "-0.01890000",
    },
    {
      period: "2013",
      mktrf: "0.29270000",
      smb: "0.06880000",
      hml: "0.01440000",
    },
    {
      period: "2012",
      mktrf: "0.14060000",
      smb: "-0.01250000",
      hml: "0.09570000",
    },
    {
      period: "2011",
      mktrf: "0.03040000",
      smb: "-0.06280000",
      hml: "-0.08700000",
    },
    {
      period: "2010",
      mktrf: "0.24840000",
      smb: "0.14740000",
      hml: "-0.05680000",
    },
    {
      period: "2009",
      mktrf: "0.23890000",
      smb: "0.09070000",
      hml: "-0.08960000",
    },
    {
      period: "2008",
      mktrf: "-0.36460000",
      smb: "0.03380000",
      hml: "0.00840000",
    },
    {
      period: "2007",
      mktrf: "-0.09380000",
      smb: "-0.06440000",
      hml: "-0.13210000",
    },
    {
      period: "2006",
      mktrf: "-0.00400000",
      smb: "0.00210000",
      hml: "0.12630000",
    },
    {
      period: "2005",
      mktrf: "0.19790000",
      smb: "-0.02220000",
      hml: "0.09600000",
    },
    {
      period: "2004",
      mktrf: "0.01690000",
      smb: "0.04500000",
      hml: "0.07500000",
    },
    {
      period: "2003",
      mktrf: "0.07480000",
      smb: "0.21780000",
      hml: "0.04490000",
    },
    {
      period: "2002",
      mktrf: "-0.36020000",
      smb: "0.04010000",
      hml: "0.06340000",
    },
    {
      period: "2001",
      mktrf: "-0.10050000",
      smb: "0.19060000",
      hml: "0.20560000",
    },
    {
      period: "2000",
      mktrf: "-0.10590000",
      smb: "-0.01610000",
      hml: "0.42420000",
    },
    {
      period: "1999",
      mktrf: "0.43110000",
      smb: "0.17380000",
      hml: "-0.37190000",
    },
    {
      period: "1998",
      mktrf: "0.12120000",
      smb: "-0.24190000",
      hml: "-0.08260000",
    },
    {
      period: "1997",
      mktrf: "0.48610000",
      smb: "-0.07880000",
      hml: "0.20350000",
    },
    {
      period: "1996",
      mktrf: "0.23010000",
      smb: "-0.03910000",
      hml: "0.06490000",
    },
    {
      period: "1995",
      mktrf: "0.20890000",
      smb: "-0.08460000",
      hml: "0.04930000",
    },
    {
      period: "1994",
      mktrf: "-0.16420000",
      smb: "-0.00960000",
      hml: "-0.00620000",
    },
  ]
  return SimpleTable(rows)
}
