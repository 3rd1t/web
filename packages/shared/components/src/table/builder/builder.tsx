import { Table } from "../table"
import { Simple } from "../cells/simple/simple"
import React from "react"
export type Row = Record<string, string | number>

/**
 * Checks for equality of two string arrays, ignoring their order.
 * @param a - string array A
 * @param b - string array B
 * @returns true if both arrays contain the same strings
 */
export const isEqual = (a: string[], b: string[]): boolean => {
  if (a.length !== b.length) {
    return false
  }
  const aSorted = a.sort()
  const bSorted = b.sort()
  for (let i = 0; i < a.length; i++) {
    if (aSorted[i] !== bSorted[i]) {
      return false
    }
  }
  return true
}

/**
 * validateFields ensure every row has the same field names.
 *
 * @param rows - Json array of fields
 * @returns True if all rows have the same field names
 */
export function validateFields(rows: Row[]): boolean {
  if (rows.length === 0) {
    return false
  }

  const firstKeys = Object.keys(rows[0])
  if (firstKeys.length === 0) {
    return false
  }

  if (rows.length === 1) {
    return true
  }
  for (let i = 1; i < rows.length; i++) {
    const compareKeys = Object.keys(rows[i])
    if (!isEqual(firstKeys, compareKeys)) {
      return false
    }
  }
  return true
}

export const SimpleTable = (rows: Row[]): React.ReactElement => {
  if (rows.length === 0) {
    return <Table columnNames={[]} cells={[]} />
  }
  if (!validateFields(rows)) {
    throw new Error("rows do not have the same field names")
  }

  const columnNames = Object.keys(rows[0])

  const simpleCells = rows.map((row) => columnNames.map((name) => <Simple label={row[name]} />))

  return <Table columnNames={columnNames} cells={simpleCells} />
}
