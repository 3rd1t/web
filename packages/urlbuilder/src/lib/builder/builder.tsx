import React, { useState } from "react"
import { AutoSuggest, Status } from "./auto-suggest/auto-suggest"

/* eslint-disable-next-line */
export interface BuilderProps {}

export const Builder = (props: BuilderProps) => {
  const newState = (initial: string | Status) => {
    const [state, setState] = useState(initial)
    return {
      state,
      setState,
    }
  }

  const fields = [
    {
      name: "factor",
      value: newState(""),
      status: newState(Status.IDLE),

      options: ["3", "4", "5", "6"],
    },
    {
      name: "region",
      value: newState(""),
      status: newState(Status.IDLE),

      options: [
        "USA",
        "Developed",
        "Developed_ex_US",
        "Europe",
        "Japan",
        "Asia_Pacific_ex_Japan",
        "North_America",
        "Emerging",
      ],
    },
    {
      name: "currency",
      value: newState(""),
      status: newState(Status.IDLE),

      options: [
        "EUR",
        "JPY",
        "GBP",
        "CHF",
        "RUB",
        "AUD",
        "BRL",
        "CAD",
        "CNY",
        "INR",
        "DKK",
        "NZD",
        "NOK",
        "SEK",
        "PLN",
        "ILS",
        "KRW",
        "TRY",
      ],
    },
    {
      name: "interval",
      value: newState(""),
      status: newState(Status.IDLE),

      options: ["daily", "monthly", "annual"],
    },
  ]

  const isValid = () => {
    const v = fields.every((field) => {
      return field.status.state === Status.DONE
    })
    console.log(v)
    return v
  }

  return (
    <div>
      <div>
        <h2></h2>
      </div>
      <form className="flex items-center justify-center mt-20 appearance-none">
        <div className="flex items-center mx-20 text-gray-700 border border-gray-300 rounded shadow">
          <span className="p-2 font-thin bg-gray-100 border-r border-gray-300">https://api.perfol.io/v1/factor/</span>

          {fields.map((f, i) => (
            <div key={i} className="flex items-center">
              {i > 0 ? <span className="font-thin"> /</span> : null}
              {
                <AutoSuggest
                  placeholder={f.name}
                  value={f.value.state}
                  options={f.options}
                  updateValue={f.value.setState}
                  status={f.status.state}
                  updateStatus={f.status.setState}
                />
              }
            </div>
          ))}
          <span className="p-2 font-thin bg-gray-100 border-l border-gray-300">
            <svg fill="none" viewBox="0 0 24 24" stroke="currentColor" className="w-6 h-6 text-green-500">
              {isValid() ? (
                <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M5 13l4 4L19 7" />
              ) : (
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth={2}
                  d="M12 9v2m0 4h.01m-6.938 4h13.856c1.54 0 2.502-1.667 1.732-3L13.732 4c-.77-1.333-2.694-1.333-3.464 0L3.34 16c-.77 1.333.192 3 1.732 3z"
                ></path>
              )}
            </svg>
          </span>
        </div>
      </form>
    </div>
  )
}

export default Builder
