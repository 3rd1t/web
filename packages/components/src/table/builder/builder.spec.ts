import { validateFields, isEqual, SimpleTable } from "./builder"

import { render } from "@testing-library/react"

describe("validateFields", () => {
  const testCases: {
    name: string
    rows: Row[]
    want: boolean
  }[] = [
    {
      name: "invalidates empty list",
      rows: [],
      want: false,
    },
    {
      name: "invalidates a single empty row",
      rows: [{}],
      want: false,
    },
    {
      name: "invalidates multiple empty rows",
      rows: [{}, {}],
      want: false,
    },
    {
      name: "invalidates if at least one row is empty",
      rows: [{ name: "Hello" }, {}],
      want: false,
    },
    {
      name: "validates with empty string",
      rows: [{ name: "Hello" }, { name: "" }],
      want: true,
    },
    {
      name: "validates a single row and column with string",
      rows: [{ name: "Hello" }],
      want: true,
    },
    {
      name: "validates a single row and column with number",
      rows: [{ age: 1 }],
      want: true,
    },
    {
      name: "validates a single column with strings",
      rows: [{ name: "Hello" }, { name: "World" }],
      want: true,
    },
    {
      name: "validates a single row with multiple strings",
      rows: [{ name: "Hello", surname: "World" }],
      want: true,
    },
    {
      name: "validates a single row with multiple strings and numbers",
      rows: [{ name: "Hello", surname: "World", age: 1 }],
      want: true,
    },
    {
      name: "validates a multiple rows with multiple strings and numbers",
      rows: [
        { name: "Hello", surname: "World", age: 1 },
        { name: "Bert", surname: "John", age: 5 },
      ],
      want: true,
    },
  ]

  testCases.forEach((tc) => {
    it(tc.name, () => {
      expect(validateFields(tc.rows)).toBe(tc.want)
    })
  })
})

describe("isEqual", () => {
  const testCases: {
    name: string
    a: string[]
    b: string[]
    want: boolean
  }[] = [
    {
      name: "invalidates inequal length",
      a: [],
      b: ["hello"],
      want: false,
    },
    {
      name: "invalidates different lists",
      a: ["world"],
      b: ["hello"],
      want: false,
    },
    {
      name: "validates a single string",
      a: ["hello"],
      b: ["hello"],
      want: true,
    },
    {
      name: "validates multiple strings",
      a: ["hello", "world"],
      b: ["hello", "world"],
      want: true,
    },
    {
      name: "validates multiple unsorted strings",
      a: ["world", "hello"],
      b: ["hello", "world"],
      want: true,
    },
  ]
  testCases.forEach((tc) => {
    it(tc.name, () => {
      expect(isEqual(tc.a, tc.b)).toBe(tc.want)
    })
  })
})

describe("SimpleTable", () => {
  const rows = [
    {
      name: "Hello",
      age: 1,
    },
    {
      name: "World",
      age: 2,
    },
    {
      name: "Bruce",
      age: 3,
    },
    {
      name: "Wayne",
      age: 4,
    },
  ]
  it("should match snapshot", () => {
    const { baseElement } = render(SimpleTable(rows))
    expect(baseElement).toMatchSnapshot()
  })
})
