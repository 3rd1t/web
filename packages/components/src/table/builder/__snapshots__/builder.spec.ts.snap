// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleTable should match snapshot 1`] = `
<body>
  <div>
    <div
      class="flex flex-col "
    >
      <div
        class="-my-2 overflow-x-auto sm:-mx-6 lg:-mx-8"
      >
        <div
          class="inline-block min-w-full py-2 align-middle sm:px-6 lg:px-8"
        >
          <div
            class="overflow-hidden border border-gray-300 sm:rounded"
          >
            <table
              class="min-w-full "
            >
              <thead>
                <tr>
                  <th
                    class="
                text-left
                px-6 py-4 text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase bg-gray-100"
                  >
                    name
                  </th>
                  <th
                    class="
                text-right
                px-6 py-4 text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase bg-gray-100"
                  >
                    age
                  </th>
                </tr>
              </thead>
              <tbody
                class="bg-white divide-y divide-gray-200"
              >
                <tr>
                  <td
                    class="text-left"
                  >
                    <div
                      class="px-6 py-4 whitespace-no-wrap"
                    >
                      <span
                        class="text-sm leading-5 text-right text-gray-900"
                      >
                        Hello
                      </span>
                    </div>
                  </td>
                  <td
                    class="text-right"
                  >
                    <div
                      class="px-6 py-4 whitespace-no-wrap"
                    >
                      <span
                        class="text-sm leading-5 text-right text-gray-900"
                      >
                        1
                      </span>
                    </div>
                  </td>
                </tr>
                <tr>
                  <td
                    class="text-left"
                  >
                    <div
                      class="px-6 py-4 whitespace-no-wrap"
                    >
                      <span
                        class="text-sm leading-5 text-right text-gray-900"
                      >
                        World
                      </span>
                    </div>
                  </td>
                  <td
                    class="text-right"
                  >
                    <div
                      class="px-6 py-4 whitespace-no-wrap"
                    >
                      <span
                        class="text-sm leading-5 text-right text-gray-900"
                      >
                        2
                      </span>
                    </div>
                  </td>
                </tr>
                <tr>
                  <td
                    class="text-left"
                  >
                    <div
                      class="px-6 py-4 whitespace-no-wrap"
                    >
                      <span
                        class="text-sm leading-5 text-right text-gray-900"
                      >
                        Bruce
                      </span>
                    </div>
                  </td>
                  <td
                    class="text-right"
                  >
                    <div
                      class="px-6 py-4 whitespace-no-wrap"
                    >
                      <span
                        class="text-sm leading-5 text-right text-gray-900"
                      >
                        3
                      </span>
                    </div>
                  </td>
                </tr>
                <tr>
                  <td
                    class="text-left"
                  >
                    <div
                      class="px-6 py-4 whitespace-no-wrap"
                    >
                      <span
                        class="text-sm leading-5 text-right text-gray-900"
                      >
                        Wayne
                      </span>
                    </div>
                  </td>
                  <td
                    class="text-right"
                  >
                    <div
                      class="px-6 py-4 whitespace-no-wrap"
                    >
                      <span
                        class="text-sm leading-5 text-right text-gray-900"
                      >
                        4
                      </span>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
